homeassistant/components/snapcast/media_player.py

```diff
@@ -22,72 +21,51 @@
     STATE_PLAYING,
     STATE_UNKNOWN,
 )
-import homeassistant.helpers.config_validation as cv
-from homeassistant.helpers.dispatcher import async_dispatcher_connect
+from homeassistant.helpers import config_validation as cv, entity_platform
 
-from . import (
+from .const import (
+    ATTR_LATENCY,
     ATTR_MASTER,
-    DOMAIN,
+    CLIENT_PREFIX,
+    CLIENT_SUFFIX,
+    DATA_KEY,
+    GROUP_PREFIX,
+    GROUP_SUFFIX,
     SERVICE_JOIN,
     SERVICE_RESTORE,
+    SERVICE_SET_LATENCY,
     SERVICE_SNAPSHOT,
     SERVICE_UNJOIN,
 )
 
 _LOGGER = logging.getLogger(__name__)
 
-DATA_KEY = "snapcast"
-
 SUPPORT_SNAPCAST_CLIENT = (
     SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET | SUPPORT_SELECT_SOURCE
 )
 SUPPORT_SNAPCAST_GROUP = (
     SUPPORT_VOLUME_MUTE | SUPPORT_VOLUME_SET | SUPPORT_SELECT_SOURCE
 )
 
-GROUP_PREFIX = "snapcast_group_"
-GROUP_SUFFIX = "Snapcast Group"
-CLIENT_PREFIX = "snapcast_client_"
-CLIENT_SUFFIX = "Snapcast Client"
-
-PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend(
-    {vol.Required(CONF_HOST): cv.string, vol.Optional(CONF_PORT): cv.port}
-)
-
 
 async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up the Snapcast platform."""
 
     host = config.get(CONF_HOST)
     port = config.get(CONF_PORT, CONTROL_PORT)
 
-    async def async_service_handle(service_event, service, data):
-        """Handle dispatched services."""
-        entity_ids = data.get(ATTR_ENTITY_ID)
-        devices = [
-            device for device in hass.data[DATA_KEY] if device.entity_id in entity_ids
-        ]
-        for device in devices:
-            if service == SERVICE_SNAPSHOT:
-                device.snapshot()
-            elif service == SERVICE_RESTORE:
-                await device.async_restore()
-            elif service == SERVICE_JOIN:
-                if isinstance(device, SnapcastClientDevice):
-                    master = [
-                        e
-                        for e in hass.data[DATA_KEY]
-                        if e.entity_id == data[ATTR_MASTER]
-                    ]
-                    if isinstance(master[0], SnapcastClientDevice):
-                        await device.async_join(master[0])
-            elif service == SERVICE_UNJOIN:
-                if isinstance(device, SnapcastClientDevice):
-                    await device.async_unjoin()
-
-        service_event.set()
-
-    async_dispatcher_connect(hass, DOMAIN, async_service_handle)
+    platform = entity_platform.current_platform.get()
+    platform.async_register_entity_service(SERVICE_SNAPSHOT, {}, "snapshot")
+    platform.async_register_entity_service(SERVICE_RESTORE, {}, "async_restore")
+    platform.async_register_entity_service(
+        SERVICE_JOIN, {vol.Required(ATTR_MASTER): cv.entity_id}, handle_async_join
+    )
+    platform.async_register_entity_service(SERVICE_UNJOIN, {}, "async_unjoin")
```

Comment:
Looks like we should guard unjoin too, and raise if the entity is not a client.

Reply:
Good catch